### PR TITLE
Add a Duration parser for configuration.

### LINKED ISSUE #2435: Unit-Parser for configuration
**Is your feature request related to a problem? Please describe.**
Create a way to parse units in configuration parameters, of limited syntax.

E.g. `1s` would parse into 1 second duration, and `10ms` would be 10 milliseconds.

**Describe the solution you'd like**
A set of methods to pull a string in a specific format with known unit-conversion.

```
class ConfigParser {
  static long getDurationInMillis(String rawValue) {...}
  static long getSizeInBytes(String rawValue) {...}
}
```

**Describe alternatives you've considered**
We could just not allow convenient unit syntax in config.

**Additional context**
See: https://github.com/open-telemetry/opentelemetry-specification/issues/1312

### PR DESCRIPTION
**Pull request by jsuereth**
- Update ConfigBuilder in `sdk/common/export` to include a `getDurationMillisProperty` method.
- Add a short + limited duration parser that simply extract that "units" string and relies
  on substring efficiency (we assume config doesn't have to be fast as it's likely read about once)
- Add tests to verify various incarnations of duration are parsed to appropriate millisecond values.

Fixes #2435

This is my first PR to OTel-Java so sending this out for early review/evaluation. Want to verify the following:

- Is this the right location for this utility?
- Am I violating unit-test best practices?
- Do you want better documentation on the format, or can I try to contribute that directly to the speicfication?

### CODE REVIEW

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java
Comment (jkwatson) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java:195]:
```suggestion

  /** Returns the TimeUnit associated with a unit string. Defaults to milliseconds. */
```

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java
Comment (jkwatson) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java:230]:
```suggestion

  /** Returns true if the given character is allowable as a unit of duration of size. */
```

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java
Comment (anuraaga) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java:164]: Thanks a lot for the help! Configuration is still in flux, and this functionality just moved to a different location. Can you put it there instead?

https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java
Comment (anuraaga) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java:183]: I guess we could just return a `Duration`

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java
Comment (anuraaga) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java:220]: Not a big difference, but I think I would search for the first non-number to avoid the additional `isDurationUnitToken` method.

File: sdk/common/src/test/java/io/opentelemetry/sdk/common/export/ConfigBuilderTest.java
Comment (anuraaga) [inline sdk/common/src/test/java/io/opentelemetry/sdk/common/export/ConfigBuilderTest.java:87]: Instead of needing comments, we could use e.g., `TimeUnit.DAYS.toMillis` in these tests too for self-description.

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java
Comment (jsuereth) [reply sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java]: Done.  Not sure if it's too heavyweight, but it is convenient.

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java
Comment (jsuereth) [reply sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java]: Done.

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java
Comment (jsuereth) [reply sdk/common/src/main/java/io/opentelemetry/sdk/common/export/ConfigBuilder.java]: I moved to that. There's a lot of "oddities" w/ i18n and codepoints here, depending on what people paste in.  LIKELY we should throw a ConfigException on most attempts to hack this value, but my brain's still not quite up to the task of remembering the problematic codepoints + encodings to try out in tests.

File: sdk/common/src/test/java/io/opentelemetry/sdk/common/export/ConfigBuilderTest.java
Comment (jsuereth) [reply sdk/common/src/test/java/io/opentelemetry/sdk/common/export/ConfigBuilderTest.java]: Done.

**Review by jkwatson — APPROVED** [review]
Seems good to me. We should probably hold off on merging until the relevant spec stuff has been approved so we don't end up supporting dead code.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java
Comment (anuraaga) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:187]: Let's just go in order of size to reduce confusion when reading this code, the optimization of one less branch for going out-of-order is just not worth it.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java
Comment (anuraaga) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:188]: Consider `switch(unitString)`, ends up as same bytecode but less `.equals` repetition

**Review by anuraaga — APPROVED** [review]
Just some very small nits, thanks!

**Review by anuraaga — COMMENTED** [review]
@jsuereth With https://github.com/open-telemetry/opentelemetry-specification/issues/1351 filed I think we can merge this. Do you mind taking a look at the remaining comments?

**Comment by jkwatson** [comment]
@jsuereth ping! I think we could get this merged, but there are some outstanding comments, and it needs a rebase.

**Comment by jsuereth** [comment]
Sorry! Fixing now.  Was swamped in gradle + too many meetings hell for the past 48 hours.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java
Comment (jsuereth) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:218]: Two points:

1. Reverse parsing is common tactic.   If you'd like I can fix up readability.  Your code finds the first non-digit, my code find the last digit.  Renaming "i" to "lastDigitIndex" should clarify that
2. Reverse-parsing is a LOT more efficient here.  Usually my while loop is a no-op, and it should iterate over far less characters.  I'd hate to swap this in case someone does try to use this where performance matters (or copy-pasta is somewhere else without understanding that nuance).

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java
Comment (jsuereth) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:188]: Fixed

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java
Comment (jsuereth) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:187]: Removed

**Comment by jsuereth** [comment]
@jkwatson Should be updated and ready to go.

**Review by anuraaga — APPROVED** [review]
Thanks1

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java
Comment (anuraaga) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:135]: Hmm this reminds me of the spec repo PR - in case you're not familiar, you need to press "Commit suggestion" to reflect it into the code, not only resolve the conversation.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java
Comment (anuraaga) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:157]: Added a commit to change visibility and start of docstring to match a recent PR

**Pull request merged**
